MH-12613 New WorkflowOperationHandler 'create-event'

krinnewitz opened the pull request:
This patch adds a new WOH to create new events from an existing one. This is useful when you have a recording with - let's say 3 - different presentations and you want to create 3 events where each event only includes one presentation. With this WOH, you would copy the original event 2 times and cut each of the 3 events differently to only include one presentation, each.

It would be nice to have this functionality in the video editor, but as a first approach, this WOH is a good trade-off between effort and usability. Also, this WOH is useful for creating test data.

This WOH also provides the possibility to select the media package elements to be copied by tags and flavors. Tags can be changed for the newly created media package elements using a "+" and "-" syntax.

A workflow to use/test the WOH is included. It allows you to specify how many copies you would like to create.

This work is sponsored by SWITCH.

Commit:
MH-12613 New WorkflowOperationHandler 'duplicate-event'

This patch adds a new WOH to create new events from an existing one.
This is useful when you have a recording with - let's  say 3 - different
presentations and you want to create 3 events where each event only
includes one presentation. With this WOH, you would copy the original
event 2 times and cut each of the 3 events differently to only include
one presentation, each.

It would be nice to have this functionality in the video editor, but as
a first approach, this WOH is a good trade-off between effort and
usability. Also, this WOH is useful for creating test data.

This WOH also provides the possibility to select the media package
elements to be copied by tags and flavors. Tags can be changed for the
newly created media package elements using a "+" and "-" syntax.

smarquard commented:
I would suggest calling this "clone-event" rather than "create-event", which leads me to think that a new event is being created, rather than a copy of an existing event.

staubesv commented:
@smarquard Agree on "create-event" being a suboptimal name and "clone-event" being better. But since we're talking about a better name...

What would you think of "duplicate-event"?

staubesv commented:
@krinnewitz I'm somewhat worried about somebody (unintentionally or by intention) starting this operation to create 100'000+ duplicates. I think we should add a limit to prevent that serious damage this operation could then do. It would also be kind of a security vulnerability in means of denial-of-service attack if a non-system-admin user could bring down the Opencast instance by creating a very large number of duplicates.

I suggest doing so by adding a configuration key (something like "max-duplicates" - naming depends on the final name of the WOH). That way, we could set this to a quite high value for 
testing workflows (generate test data) and to a more restrictive value for production workflows. 

Besides the actual WOH, the configuration panel should also limit the number of duplicates a user can enter. That should happen in a user-friendly way, i.e. if it is possible to enter a value higher then max-duplicates, there needs to be a error message telling the user that she or he cannot enter a value higher than max-duplicates here.

krinnewitz commented:
Ok. I renamed the WOH to 'duplicate-event' and added a 'max-events' configuration value. I did not rename the git branch, nor the title of this PR since that would mean that I have to re-create the whole PR. If you guys think this is necessary, please let me know.

lkiesow commented on etc/workflows/duplicate-event.xml:
```diff
@@ -0,0 +1,73 @@
+        </fieldset>
+      </div>
+
+      <script type="text/javascript">
```
Just wondering why you added this?

- The `min` and `max` attributes should be used by all modern browsers and should protect users from entering weird values, shouldn't they?
- It's not providing any security since users can change the values if they really want to do stupid stuff anyway (the code is executed in the browser) and the backend will check the values anyway
- It makes the workflow more complex since you require users to understand JS

staubesv replied:
I've asked Kim to ensure that users cannot enter numbers that won't work ;-)

The workflow operation handler itself mitigates a denial-of-service attack by limiting the maximum number of duplicates it will create. This implies, however, that the front-end does this limitation, too: Otherwise, users will wonder why the workflow did not create as many duplicates as they have asked for. So this is a usability feature - not a security feature.

If this can be done easier (min and max), that is of course preferable.

krinnewitz replied:
min and max only apply when the user uses the up and down arrows to set the value of the number field. He can also enter the number freely and in this case, he is able to enter negative numbers as well as numbers higher than max. At least with the browser I used for testing (I think it was chrome). 

The javascript code suppresses the "-" sign as well as numbers which exceed min/max.

Of course, this has nothing to do with security, but is a usability feature.

I think this workflow definition is just an example. Admins may customize it to fit the needs of their users.

lkiesow replied:
Ok, makes sense. Thanks.

> I think this workflow definition is just an example. Admins may customize it to fit the needs of their users.

Examples should go into the documentation, not in the default set of workflows. But I think it's ok to have this as default workflow. Though maybe we should drop max to something more reasonable like 25?

lkiesow commented on etc/workflows/duplicate-event.xml:
```diff
@@ -0,0 +1,73 @@
+    <operation
+      id="defaults"
+      description="Applying default configuration values">
+      <configurations>
```
It's a bit weird to have the default values for a single operation in the global workflow configuration. Wouldn't it make more sense to add defaults to the operation itself?

lkiesow commented on etc/workflows/duplicate-event.xml:
```diff
@@ -0,0 +1,73 @@
+      exception-handler-workflow="partial-error"
+      description="Duplicate Event">
+      <configurations>
+        <configuration key="source-tags">archive</configuration>
```
The tag is redundant, isn't it?

krinnewitz replied:
@staubesv what was the idea behind the requirement to make the source tags configurable? Do we need  this or shall it rather be hard-coded to "archive"?

staubesv replied:
@krinnewitz The idea is to make the workflow operation more flexible considering the media package elements that the duplicate will contain. Besides that, having the source tag specified explicitly is makes it more obvious what the WOH does

@lkiesow The tag is not redundant as the duplicate is not a clone - it is rather a new event that contains the media package elements of the currently processed media package identified by source-tags (plus, admittedly, some less obvious things as the internal publication and a modified dublin core catalog)

staubesv replied:
@krinnewitz & @lkiesow Oh, just see what Lars means with redundant: If you have source-flavors=\*/\*, that will already select all media package elements so there is indeed no need for source-tags at all. Please also adjust the example in the documentation. On our setup, we will likely do a "source-tags=archive" as we don't want to select all media package elements, but for the example workflow definition, selecting all media package elements is likely fine.

lkiesow commented on modules/workflow-workflowoperation/src/test/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandlerTest.java:
```diff
@@ -0,0 +1,323 @@
+    configurations.put(SOURCE_TAGS_PROPERTY, "archive");
+    configurations.put(TARGET_TAGS_PROPERTY, "tag1,tag2");
+    configurations.put(NUMBER_PROPERTY, "" + 1);
+    configurations.put(MAX_NUMBER_PROPERTY, "" + 10);
```
Please wrap lines after 120 characters. This goes for several lines below as well

lkiesow commented on modules/workflow-workflowoperation/src/test/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandlerTest.java:
```diff
@@ -0,0 +1,323 @@
+
+
+    Capture<MediaPackage> mpCapt = Capture.newInstance();
+    expect(ingestService.addCatalog(anyObject(), anyString(), eq(MediaPackageElements.EPISODE), capture(mpCapt))).andAnswer(
```
Why not use:

```Java
expect(ingestService.addCatalog(anyObject(), anyString(), eq(MediaPackageElements.EPISODE), capture(mpCapt)))
    .andAnswer(mpCapt::getValue).times(numberOfCopies);
```

krinnewitz replied:
@lkiesow is there a policiy regarding the usage of java8 features in opencast? I would love to use streams, lambdas, CompletableFutures and so on, but I thought that there is some kind of rejection in the community. Hence, I avoided to use Java8 features. Am I free to use them?

staubesv replied:
@krinnewitz There is no such policy and Opencast 4.x and higher officially require Java 8. I don't think there is any reason why you would need to avoid Java 8 features... feel free to use them.

staubesv replied:
@krinnewitz ... but there is a code style guide saying "no more than 120 characters per line", see https://docs.opencast.org/develop/developer/code-style/

krinnewitz replied:
Yes, alright alright, I am already fixing that :D This maybe something that could go into the checksytle rules.

krinnewitz replied:
hmm, looks like our version of felix doesn't work with java8 features which are not byte code compatible with java7. During build, I get:

Error building bundle org.opencastproject:opencast-workflow-workflowoperation:bundle:5-SNAPSHOT : Invalid class file: org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.class

Maybe we have to update maven plugins for this to work. I won't do it in this PR, though.

lkiesow commented on modules/workflow-workflowoperation/src/test/resources/duplicate-event_mediapackage.xml:
```diff
@@ -0,0 +1,116 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<ns2:mediapackage xmlns:ns2="http://mediapackage.opencastproject.org" start="2010-05-19T09:27:46" id="d51f9d4e-b5e7-41e0-8640-0d1f43b2fed0" duration="65649">
```
Not that important since it's a test resource but indentation should usually be 2 spaces

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,467 @@
+  private static final String MINUS = "-";
+
+  /** The configuration options for this handler */
+  private static final SortedMap<String, String> CONFIG_OPTIONS;
```
What do you use these for? While these are still in some other WOH, iirc these are completely obsolete and you do not need them. As far as I can see, you never use them either.

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,467 @@
+  /** The ingest service */
+  private IngestService ingestService;
+
+  static {
```
again: obsolete

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,467 @@
+    final MediaPackage mediaPackage = workflowInstance.getMediaPackage();
+    final WorkflowOperationInstance currentOperation = workflowInstance.getCurrentOperation();
+
+    final String configuredSourceFlavors = StringUtils.trimToEmpty(currentOperation.getConfiguration(SOURCE_FLAVORS_PROPERTY));
```
line length (>120 chars)

lkiesow reviewed (changes requested):
Here are a few minor things to fix. A review of the main source code file will follow later today or tomorrow morning.

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,467 @@
+      elementSelector.addFlavor(MediaPackageElementFlavor.parseFlavor(flavor));
+    }
+
+    final List<String> removeTags = new ArrayList<String>();
```
Just a minor remark that with Java8 `… = new ArrayList<>();` would work as well. The `<String>` is redundant.

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,467 @@
+    final Collection<Publication> internalPublications = new HashSet<>();
+
+    for (MediaPackageElement e: mediaPackage.getElements()) {
+      if (e instanceof Publication && CHANNEL_ID.equals(((Publication) e).getChannel())) {
```
I think using `InternalPublicationChannel.CHANNEL_ID` would actually be less confusing here.

Though I'm also wondering why you would copy any publications for a new media package at all. Does this not have any side effects? What if you remove the publication for the original event (e.g. by deleting it).?

staubesv replied:
The motivation to copy the internal publication is that we expect this operation to be used the following way:

1. Record a series of n short talks
2. Create n-1 duplicates
3. For each duplicate, open video editor to edit metadata and cut out one talk

If we would not "copy" the internal publication, a user would need to re-render the video editor previews for each of the videos which is very undesirable as this takes quite a while.

Not sure whether it is implemented that way, but the WOH should do a deep copy of the internal publication (incl. distribution of the preview tracks for the individual media packages) to avoid hidden dependencies of duplicates. We don't expect the file duplication to be a problem as this kind of recording is not expected to happen often.

@krinnewitz It does do that deep-copy thing, doesn't it?

krinnewitz replied:
@staubesv yes, it should copy all the related files.

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,467 @@
+    if (episodeDcs.length == 1) {
+      originalDublinCore = Opt.some(episodeDcs[0]);
+    } else {
+      logger.warn("Media package {} has {} episode dublin cores while it is expected to have exactly 1.",
```
Should it not do more than warn here? E.g. if there are two (shouldn't usually happen) then you would still have removed them earlier and you would end with none …which should cause an error later on. In that case, just failing here would probably be best.

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,467 @@
+        final MediaPackageElement element = (MediaPackageElement) e.clone();
+        final File sourceFile = workspace.get(element.getURI());
+        final URI tmpUri = workspace.put(
+                destination.getIdentifier().toString(), element.getIdentifier().toString(),
```
`element.getIdentifier()` already returns a `String`

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,467 @@
+
+        // Now we can distribute it to the new media package
+        destination.add(element); // Element has to be added before it can be distributed
+        final Job j = distributionService.distribute(CHANNEL_ID, destination, element.getIdentifier());
```
Please don't call your variables `j`. Call it `job`.

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,467 @@
+    return destination;
+  }
+
+  private void copyPublication(
```
Three questions about this:

1. Why do you copy publications at all? It seems weird and not something you would expect an event duplication woh to do at all. If you want to publish it, why not publish it the regular way?
2. Why copy files into the workspace and the working file repository at all? Why not operate on the existing files?
3. You never remove any of the files you temporarily put into the workspace and the working file repository: You leave a lot of trash lying around on the filesystem.

krinnewitz replied:
Only the internal publication is copied. It is needed for the video editor to work. If I don't copy it, the duplicated event cannot be edited. As far as I remember, I am putting the files into the workspace in order create a new URI. I did not find another way to do this without hardcoding any URI patterns. Does that make sense? I will add the cleanup operation to avoid files being left in the workspace.

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,467 @@
+    // Using an altered copy of the source publication's URI is a bit hacky, but it works without knowing the URI pattern...
+    String publicationUri = sourcePublication.getURI().toString();
+    publicationUri = publicationUri.replace(source.getIdentifier().toString(), destination.getIdentifier().toString());
+    publicationUri = publicationUri.replace(sourcePublication.getIdentifier().toString(), newPublicationId);
```
`sourcePublication.getIdentifier()` returns a `String`

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,467 @@
+
+  private MediaPackage copyMediaPackage(MediaPackage source, long copyNumber, String copyNumberPrefix) throws WorkflowOperationException {
+    // We are not using MediaPackage.clone() here, since it does "too much" for us (e.g. copies all the attachments)
+    MediaPackage destination;
```
Wouldn't `source.clone()` + replacing the media package identifier do here? It would also be safer in case someone modifies the media package structure in the future.

krinnewitz replied:
I think source.clone() would do a much "deeper" copy and add a lot of stuff we don't want, like attachments and so on and I would have to remove all these things. Or am I wrong?

lkiesow commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,467 @@
+    }
+    updateTags(destinationDublinCore, removeTags, addTags, overrideTags);
+    try (InputStream inputStream = IOUtils.toInputStream(destinationDublinCore.toXmlString(), "UTF-8")) {
+      destination = ingestService.addCatalog(inputStream, "dublincore.xml", MediaPackageElements.EPISODE, destination);
```
I'm wondering why you use the ingest service to add a file to this media package and not just put the file into the ws/wfr add it directly? You add a dependency to another module for only this single line.

krinnewitz replied:
That's a good point. I mostly try to avoid redundancy but in this case, these are so few lines that it makes sense to drop the dependency instead.

Can you say something about Workspace vs. WorkingFileRepository? They look very similar. Is it sufficient if I use the Workspace here? The IngestService uses WorkingFileRepository, but so far, I only have the workspace as a dependency in the WOH.

lkiesow commented on etc/workflows/duplicate-event.xml:
```diff
@@ -0,0 +1,73 @@
+    </operation>
+
+  </operations>
+
```
This one should also get a `cleanup` operation. See general comment for more details.

lkiesow reviewed (dismissed):
The usage of assetmanager properties is a bit weird. For once, there is an error if there was no previous snapshot which is totally unnecessary since this is basically the snapshot operation with the slight difference that you cange the identifier and title:

    curl -f -i --digest \
      -u opencast_system_account:CHANGE_ME \
      -H "X-Requested-Auth: Digest" \
      http://localhost:8080/ingest/addMediaPackage/duplicate-event \
      -F 'flavor=presentation/source' -F 'BODY=@test.mp4' \
      -F title="`uuidgen`" -F creator="`uuidgen`"

…results in…

    2018-01-18 16:07:23,697 | ERROR | (DuplicateEventWorkflowOperationHandler:433) - AssetManager could not find source media package f2a9543b-b3e0-4203-8658-54eb6f0ded4f. This should never happen.
    2018-01-18 16:07:23,699 | ERROR | (DuplicateEventWorkflowOperationHandler:433) - AssetManager could not find source media package f2a9543b-b3e0-4203-8658-54eb6f0ded4f. This should never happen.

While this is declared an error, the operation continues and claims success which is a bit weird. It should fail if it is an error or not log an error. This will also make you end up with all events having the same title. Probably because it cannot get the properties:

    sometitle (copy 1)
    sometitle (copy 1)
    sometitle (copy 1)
    …

Why not simply count internally? You cannot guarantee that titles are unique anyway since users can change them. In fact if you duplicate an event to edit and process a part you will probably rename it as well. If you then later come back and want to extract a different part, it's even wird to start with `copy 2`. It just sounds simpler and less error prone.

---

In general, you need to clean up after your operation. This is how my workspace and working file repository looks after running your workflow:

    files/mediapackage/7798e3fd-e3a8-4e53-9db3-5145ab6ec57b/d64af8bb-e256-4df5-951e-25bc11dcfb8d/dublincore.xml
    files/mediapackage/7798e3fd-e3a8-4e53-9db3-5145ab6ec57b/d64af8bb-e256-4df5-951e-25bc11dcfb8d/dublincore.xml.md5
    files/mediapackage/f2a9543b-b3e0-4203-8658-54eb6f0ded4f/security-policy-series/xacml.xml.md5
    files/mediapackage/f2a9543b-b3e0-4203-8658-54eb6f0ded4f/security-policy-series/xacml.xml
    files/mediapackage/f2a9543b-b3e0-4203-8658-54eb6f0ded4f/37e8bd11-343f-4475-ac4c-a3d8aad43cb4/test.mp4
    files/mediapackage/f2a9543b-b3e0-4203-8658-54eb6f0ded4f/37e8bd11-343f-4475-ac4c-a3d8aad43cb4/test.mp4.md5
    files/mediapackage/f2a9543b-b3e0-4203-8658-54eb6f0ded4f/71f7a2f8-c29d-4304-835c-0963a77bdf58/dublincore.xml
    files/mediapackage/f2a9543b-b3e0-4203-8658-54eb6f0ded4f/71f7a2f8-c29d-4304-835c-0963a77bdf58/dublincore.xml.md5
    workspace/mediapackage/7798e3fd-e3a8-4e53-9db3-5145ab6ec57b/d64af8bb-e256-4df5-951e-25bc11dcfb8d/dublincore.xml
    workspace/mediapackage/f2a9543b-b3e0-4203-8658-54eb6f0ded4f/security-policy-series/xacml.xml
    workspace/http_localhost_8080/assets/assets/7798e3fd-e3a8-4e53-9db3-5145ab6ec57b/d64af8bb-e256-4df5-951e-25bc11dcfb8d/0/dublincore.xml

You should ensure that you delete temporary files you put into the ws/wfr and probably also add a cleanup operation to the workflow definition.

krinnewitz commented:
Thanks for the review, @lkiesow . I will look into it tomorrow.

lkiesow commented on modules/workflow-workflowoperation/pom.xml:
```diff
@@ -133,6 +144,7 @@
               OSGI-INF/operations/cleanup.xml,
               OSGI-INF/operations/configure-by-dcterm.xml,
               OSGI-INF/operations/copy.xml,
+              OSGI-INF/operations/duplicate-event.xml,
```
Can you push this a line down to have the list in alphabetical order?

staubesv commented on docs/guides/admin/docs/workflowoperationhandlers/duplicate-event-woh.md:
```diff
@@ -0,0 +1,48 @@
+The duplicate event operation can be used to duplicate an event by copying an existing one. The main use case are events,
+which contain a series of different presentations which were all recorded with just one recording. In order to create
+seperate events for each presentation, the original recording can be copied and each copy can be cut to only contain
+one presentation. This operation only works with events, which were not published, yet.
```
Why should this only work with events that are not published yet?

Please add some text that describes the handling of existing publications.

krinnewitz commented:
Okay, I fixed the review findings. Some notes:

- I removed the code which persistently stored the duplicate count. I think @lkiesow is right. This is a lot of foo code which doesn't bring much benefit.
- I am now cleaning up the workspace, but I am not sure if this is sufficient. The workspace seems to also store files in the working file repository and those files there seem to not be cleaned up.

lkiesow commented on modules/workflow-workflowoperation/src/test/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandlerTest.java:
```diff
@@ -0,0 +1,296 @@
+    reset(workspace, assetManager, distributionService);
+
+    URI uriDc = getClass().getResource("/dublincore.xml").toURI();
+    expect(workspace.read(eq(URI.create("dublincore.xml")))).andReturn(new File(uriDc)).times(numberOfCopies);
```
This line conflicts with an earlier merged patch since `Workspace::read(…)` now returns an `InputStream` which causes a build to fail after it is merged into `develop`.

krinnewitz replied:
Thanks, this should be fixed now

staubesv commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,404 @@
+          inputStream);
+      final MediaPackageElement mpe = destination.add(newUrl, MediaPackageElement.Type.Catalog,
+          MediaPackageElements.EPISODE);
+      mpe.setIdentifier(elementId);
```
@krinnewitz In our older version of this code we have found a bug which, as far as I can tell, is still present in the new version: The copied dublincore/episode catalog won't have any tags, will it? 
While you correctly set them for "destinationDublinCore", they are never set to "mpe". 

Could you please check this?

krinnewitz replied:
Yes, thank you. This is now fixed.

lkiesow commented:
Hi Kim,
I totally forgot about this but finally had another look today. It looks mostly good except that you still leave files in the workspace and working file repository, A short background:

- The workspace may be local to Opencast nodes and you should always clean up after you are done with an operation.
- The working file repository is the central space to store stuff you need during a workflow run. That is why you would leave files there at the end of an operation. A workflow's final cleanup operation will take care of it.

But in this operation you are doing something unusual: You create media packages ouside the context of the workflow since you throw them into the asset manager. For that you also put data into the workspace and the working file repository. Since no one, except your operation, knows about these data, no one will remove them: You have to do that yourself.

Here, you copy the distribution artifacts but leave them in the workspace and the new dublincore catalogs. You should remove both before ending this operation

For the dublincore catalogs, something like this should suffice (untested):

```diff
diff --git a/modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java b/modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java
index b7909d218..d51c4ab5c 100644
--- a/modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java
+++ b/modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java
@@ -252,6 +252,8 @@ public class DuplicateEventWorkflowOperationHandler extends AbstractWorkflowOper
 
       // Remove temporary files of new media package
       try {
+        final String episodeDCCatalog = newMp.getCatalogs(MediaPackageElements.EPISODE)[0].getIdentifier();
+        workspace.delete(newMp, episodeDCCatalog);
         workspace.cleanup(newMp.getIdentifier());
       } catch (IOException e) {
         logger.warn("Failed to cleanup the workspace for media package {}", newMp.getIdentifier());
```

You would need to do something similar for the distribution artifacts you duplicate (though I'm still not sure why they would need to be duplicated at all…)

krinnewitz commented:
Hey Lars, thanks for your comments. I added some code to remove these files. However, one file still stays in the workspace: 

`workspace/http_localhost_8080/assets/assets/9c37eeb6-bea8-4cea-af11-89bc4a2e5994/7125d6d2-75ae-4362-814b-e0e3fab3e4ca/0/xacml.xml`

This file is created by the asset manager and I am not sure if this is something to worry about. Do you think I should delete it as well? 

In this case, I might have to add some hacky code since the media package only carries version 1 instead of version 0 of this element...

lkiesow commented:
No hack ;-P

Unfortunately, that's broken code elsewhere.
I'll try to have this tested until Monday

lkiesow commented:
Sorry to be annoying but I took a short look at your new patch and it introduces a huge concurrency problem: You introduce a global variable which is possibly used by several operations at once and hence randomly reset. Additionally, operations would delete files of other operations if they are executed simultaneously, leading to arbitrary IO exceptions.

krinnewitz commented:
No problem. For some reason I forgot about the fact that the WOH is a service and there will be a shared instance when running concurrently. I changed the variable declaration to be method-local. That should be safe now ;)

lkiesow commented:
Hi Kim,
there are a few remaining issues for which I simply provided a patch:

> https://github.com/lkiesow/opencast/commit/3b51cd11891d8716a4127d8f2be9331f010ac5a3

In short this:

- Fixes a checkstyle issue
- Fixes the dependency declaration
- Fixes leaving files in the workspace (you forgot one)

Can you check if that patch is fine for you and include it? Since this pull request is now split across a lot of commits, would you mind squashing them together (feel free to squash my patch as well).

krinnewitz commented:
Thanks @lkiesow 

I applied your patch and squashed everything into just one commit.

lkiesow commented:
merged!
